### ISSUE TITLE
Optimize Peer Joining Messages

### ISSUE THREAD

**Issue by clarkmcc** [netbirdio/netbird#1303]
As [documented here](https://netbirdio.slack.com/archives/C028VPB34NB/p1691336732336609), we have an arrangement between administrators and peers, where there may be 1-5 administrators and 10,000+ peers. As I understand it, some messages like "peer joining" messages must be distributed to all the peers in a mesh network, however, our mesh according to the ACLs do not allow for "peer" to "peer" communication, therefore it would be wasteful to distribute peer joining messages to all the "peers" rather than just the administrators.

**Comment by nazarewk** [comment]
@clarkmcc Is this still relevant for you? Could you elaborate?

I am closing this for lack of activity, but will surely reopen if you get back with more information :)

**Comment by clarkmcc** [comment]
Yes, this is still relevant for us. In a large mesh, distributing irrelevant messages to all the peers as frequently as those messages would be sent in a large mesh would be wasteful. But I can understand that maybe we have a niche use-case.

**Comment by nazarewk** [comment]
I'm not sure what you think/mean by "administrator" and "peers" distinction, but in NetBird, all Peers are equal. They distribute messages only within the scope of pairs of Peers maintaining a single connection. This applies equally to the "peer joining" and any other messages, there are no globally distributed involved anywhere in the process.

There are 2-3 caveats/exceptions to the above:
- centralized management service distributes "network updated" packets to the relevant & currently connected peers,
- it is relatively easy to establish a full-mesh with `All` group on both sides of the ACL policy,
- it is relatively easy to make a single Peer establish a connectivity with the rest of the network by using `All` group on one side of the ACL policy,

Periodical "keepalive" packets are required to determine the state of the so-called "connections". Wireguard uses UDP as transport. Then, UDP concerns itself with the scope of a single packet, which does not guarantee any kind of stable or persistent connection (unlike TCP). Increasing the frequency of those packets correlates directly with increased robustness and quicker recovery upon connectivity drops.

The above, combined with a lot of consumer & enterprise-grade stateful firewalls interfering with and/or forgetting UDP packets too far apart from each other. Makes the "chatter" essential to the stability of the connection between NetBird Peers (or any WG-based VPN).

Having that out of the way, we are actively working on a "lazy connections" feature (see https://github.com/netbirdio/netbird/pull/3675), which will make the Peers establish connectivity upon first use instead of maintaining them across whole lifetime of the Peer. This is probably highly relevant to your use case.


Does this answer your doubts or do you still have some ideas to improve the current state?

**Comment by clarkmcc** [comment]
Sure, let me clarify. I understand that all peers are equal in a mesh arrangement like Netbird, but it's not uncommon to impose a hub-and-spoke architecture on top of a mesh using ACLs. In our specific case, we might have 1-5 peers that are administrators and 10,000+ peers that are IoT devices. We don't want the IoT devices to talk to each other, but we do want the administrators to be able to talk to the devices. 

We can express this relationship using ACLs, but at the time I opened this ticket, I found that IoT devices were receiving peer joining messages for other IoT devices. The ACLs would indicate that the administrators should see every IoT device's peer joining message, but because the ACLs prevented IoT devices from talking to other IoT devices, it seems like distributing 10,000 peer joining messages to 10,000 peers when the 10,001th peer joined was unnecessary.
